### DISCUSSION TITLE
How do you detect when button is released but cursor is away from button?

### DISCUSSION THREAD

**Discussion by Elbon-Eastmage** [AvaloniaUI/Avalonia#14502] [Q&A] [answered]
At the moment, I'm trying to get my program to recognize when the user pressed a button but moved the cursor away before releasing it. Unfortunately, I keep running into dead ends. I did try using the OnPointerReleased event handler:

```
protected override void OnPointerReleased(PointerReleasedEventArgs e)
{
    if (e.Source is Button)
        Debug.WriteLine("Button released");

    base.OnPointerReleased(e);
}
```
but that didn't detect when the button was released.

Any ideas on how to get past this roadblock?

**Comment by maxkatz6** [comment]
You should also handle other pointer events, like Exited and CaptureLost

**Comment by stevemonaco** [comment, marked as answer]
Putting together a demo app should help you figure out the approach. It's possible you can create an attached behavior instead of subclassing. Be very mindful of the RoutingStrategies used for each event. May need to follow this with breakpoints as some events (PointerExited) are not called until the pointer is released and is followed by another event being fired immediately (PointerCaptureLost).

```xml
<Grid>
    <StackPanel HorizontalAlignment="Center" VerticalAlignment="Center">
        <TextBlock x:Name="textBlock" Text="Not Clicked" />
        <Button
            x:Name="button"
            Command="{Binding DoTheThing}"
            Content="Click" />
    </StackPanel>
</Grid>
```
```cs
public MainWindow()
{
    InitializeComponent();

    button.AddHandler(PointerPressedEvent, Button_OnPointerPressed, RoutingStrategies.Tunnel);
    button.AddHandler(PointerMovedEvent, Button_OnPointerMoved, RoutingStrategies.Tunnel);
    button.AddHandler(PointerCaptureLostEvent, Button_OnPointerCaptureLost, RoutingStrategies.Direct);
    button.AddHandler(PointerExitedEvent, Button_OnPointerExited, RoutingStrategies.Direct);
}

private void Button_OnPointerMoved(object? sender, PointerEventArgs e)
{
    if (sender is Button { IsPressed: true } button)
    {
        var pos = e.GetPosition(button);
        if (new Rect(0, 0, button.Bounds.Width, button.Bounds.Height).Contains(pos))
            textBlock.Text = "Inside";
        else
            textBlock.Text = "Outside";
    }
}

private void Button_OnPointerPressed(object? sender, PointerPressedEventArgs e)
{
    textBlock.Text = "Pressed";
}

private void Button_OnPointerCaptureLost(object? sender, PointerCaptureLostEventArgs e)
{
    textBlock.Text = "Capture Lost";
}

private void Button_OnPointerExited(object? sender, PointerEventArgs e)
{
    if (sender is Button { IsPressed: true })
        textBlock.Text = "Exited";
}
```

**Comment by Elbon-Eastmage** [comment]
Oh my goodness, thanks so much for the suggestions! I implemented a PointerCaptureLost event handler for my custom control:
```
void OnPointerCaptureLost(object sender, PointerCaptureLostEventArgs e)
{
    if (!HasOpenDialog)
        topLevel?.Focus();
}
```
and it works great! If the user clicks on the file browser button but moves the cursor away before release, a file dialog will not open, but this event handler will still run. Since HasOpenDialog is false, the focus will be shifted away from this control to the top level, triggering the data validation I wanted to happen.

Again, thanks so much. It feels great to finally be done with this problem!
